middleware.go

```diff
@@ -0,0 +1,111 @@
+package tigerblood
+
+import (
+	log "github.com/Sirupsen/logrus"
+	"go.mozilla.org/mozlogrus"
+	"github.com/DataDog/datadog-go/statsd"
+	"net/http"
+	"context"
+	"time"
+)
+
+func init() {
+	mozlogrus.Enable("tigerblood")
+	log.Printf("loaded middleware")
+}
+
+// Middleware wraps an http.Handler with additional functionality.
+type Middleware func(http.Handler) http.Handler
+
+// Run the request through all middleware
+func HandleWithMiddleware(h http.Handler, adapters []Middleware) http.Handler {
+	// To make the middleware run in the order in which they are specified,
+	// we reverse through them in the Middleware function, rather than just
+	// ranging over them
+	for i := len(adapters) - 1; i >= 0; i-- {
+		h = adapters[i](h)
+	}
+	return h
+}
+
+// addToContext add the given key value pair to the given request's context
+func addtoContext(r *http.Request, key string, value interface{}) *http.Request {
+	ctx := r.Context()
+	return r.WithContext(context.WithValue(ctx, key, value))
+}
+
+func AddDB(db *DB) Middleware {
+	return func(h http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			h.ServeHTTP(w, addtoContext(r, ctxDBKey, db))
+		})
+	}
+}
+
+func AddStatsdClient(statsdClient *statsd.Client) Middleware {
+	return func(h http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			h.ServeHTTP(w, addtoContext(r, ctxStatsdKey, statsdClient))
+		})
+	}
+}
+
+func AddViolations(violationPenalties map[string]uint) Middleware {
+	return func(h http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			h.ServeHTTP(w, addtoContext(r, ctxPenaltiesKey, violationPenalties))
+		})
+	}
+}
+
+func SetResponseHeaders() Middleware {
+	return func(h http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			w.Header().Add("Strict-Transport-Security", "max-age=31536000;") // APP-HSTS
+			w.Header().Add("Public-Key-Pins", `max-age=5184000; pin-sha256="WoiWRyIOVNa9ihaBciRSC7XHjliYS9VwUGOIud4PB18="; pin-sha256="r/mIkG3eEpVdm+u/ko/cwxzOMo1bk4TyHIlByibiA5E="; pin-sha256="YLh1dUR9y6Kja30RrAn7JKnbQG/uEtLMkBgFF2Fuihg="; pin-sha256="sRHdihwgkaib1P1gxX8HFszlD+7/gTfNvuAybgLPNis=";`) // APP-HPKP
```

Comment:
Now that I see it here, I wonder if HPKP is really useful. Hawk means browsers can't connect to this, and API clients won't keep a PKP state, so maybe we save the extra bytes and remove that header?